diff --git a/hal/Android.mk b/hal/Android.mk
index 3d50a25..258c6a8 100644
--- a/hal/Android.mk
+++ b/hal/Android.mk
@@ -91,6 +91,10 @@ ifeq ($(strip $(AUDIO_FEATURE_SAMSUNG_DUAL_SIM)),true)
 endif
 endif
 
+ifeq ($(strip $(AUDIO_FEATURE_ASUS_PADSTATION)),true)
+    LOCAL_CFLAGS += -DASUS_PADSTATION
+endif
+
 ifneq ($(filter apq8084 msm8974 msm8226 msm8610,$(TARGET_BOARD_PLATFORM)),)
 ifneq ($(strip $(AUDIO_FEATURE_ENABLED_COMPRESS_VOIP)),false)
     LOCAL_CFLAGS += -DCOMPRESS_VOIP_ENABLED
diff --git a/hal/msm8974/platform.c b/hal/msm8974/platform.c
index 0d56ff7..4d914e6 100644
--- a/hal/msm8974/platform.c
+++ b/hal/msm8974/platform.c
@@ -232,9 +232,15 @@ static char * device_table[SND_DEVICE_MAX] = {
     /* Playback sound devices */
     [SND_DEVICE_OUT_HANDSET] = "handset",
     [SND_DEVICE_OUT_SPEAKER] = "speaker",
+#ifdef ASUS_PADSTATION
+    [SND_DEVICE_OUT_PAD_SPEAKER] = "pad-speaker",
+#endif
     [SND_DEVICE_OUT_SPEAKER_REVERSE] = "speaker-reverse",
     [SND_DEVICE_OUT_HEADPHONES] = "headphones",
     [SND_DEVICE_OUT_SPEAKER_AND_HEADPHONES] = "speaker-and-headphones",
+#ifdef ASUS_PADSTATION
+    [SND_DEVICE_OUT_PAD_SPEAKER_AND_HEADPHONES] = "speaker-and-headphones",
+#endif
     [SND_DEVICE_OUT_VOICE_HANDSET] = "voice-handset",
     [SND_DEVICE_OUT_VOICE_SPEAKER] = "voice-speaker",
     [SND_DEVICE_OUT_VOICE_HEADPHONES] = "voice-headphones",
@@ -314,9 +320,15 @@ static int acdb_device_table[SND_DEVICE_MAX] = {
     [SND_DEVICE_NONE] = -1,
     [SND_DEVICE_OUT_HANDSET] = 7,
     [SND_DEVICE_OUT_SPEAKER] = 14,
+#ifdef ASUS_PADSTATION
+    [SND_DEVICE_OUT_PAD_SPEAKER] = 202,
+#endif
     [SND_DEVICE_OUT_SPEAKER_REVERSE] = 14,
     [SND_DEVICE_OUT_HEADPHONES] = 10,
     [SND_DEVICE_OUT_SPEAKER_AND_HEADPHONES] = 10,
+#ifdef ASUS_PADSTATION
+    [SND_DEVICE_OUT_PAD_SPEAKER_AND_HEADPHONES] = 202, //Quarx: Recheck
+#endif
     [SND_DEVICE_OUT_VOICE_HANDSET] = 7,
     [SND_DEVICE_OUT_VOICE_SPEAKER] = 14,
     [SND_DEVICE_OUT_VOICE_HEADPHONES] = 10,
@@ -395,8 +407,14 @@ struct name_to_index {
 static struct name_to_index snd_device_name_index[SND_DEVICE_MAX] = {
     {TO_NAME_INDEX(SND_DEVICE_OUT_HANDSET)},
     {TO_NAME_INDEX(SND_DEVICE_OUT_SPEAKER)},
+#ifdef ASUS_PADSTATION
+    {TO_NAME_INDEX(SND_DEVICE_OUT_PAD_SPEAKER)},
+#endif
     {TO_NAME_INDEX(SND_DEVICE_OUT_SPEAKER_REVERSE)},
     {TO_NAME_INDEX(SND_DEVICE_OUT_HEADPHONES)},
+#ifdef ASUS_PADSTATION
+    {TO_NAME_INDEX(SND_DEVICE_OUT_PAD_SPEAKER_AND_HEADPHONES)},
+#endif
     {TO_NAME_INDEX(SND_DEVICE_OUT_SPEAKER_AND_HEADPHONES)},
     {TO_NAME_INDEX(SND_DEVICE_OUT_VOICE_HANDSET)},
     {TO_NAME_INDEX(SND_DEVICE_OUT_VOICE_SPEAKER)},
@@ -1535,7 +1553,11 @@ snd_device_t platform_get_output_snd_device(void *platform, audio_devices_t devi
             }
         } else if (devices & AUDIO_DEVICE_OUT_ANLG_DOCK_HEADSET ||
                    devices & AUDIO_DEVICE_OUT_DGTL_DOCK_HEADSET) {
+#ifndef ASUS_PADSTATION
             snd_device = SND_DEVICE_OUT_USB_HEADSET;
+#else
+            snd_device = SND_DEVICE_OUT_PAD_SPEAKER;
+#endif
 #ifdef FM_ENABLED
         } else if (devices & AUDIO_DEVICE_OUT_FM_TX) {
             snd_device = SND_DEVICE_OUT_TRANSMISSION_FM;
@@ -1578,12 +1600,20 @@ snd_device_t platform_get_output_snd_device(void *platform, audio_devices_t devi
         else
             snd_device = SND_DEVICE_OUT_BT_SCO;
     } else if (devices & AUDIO_DEVICE_OUT_AUX_DIGITAL) {
-        snd_device = SND_DEVICE_OUT_HDMI ;
-    } else if (devices & AUDIO_DEVICE_OUT_ANLG_DOCK_HEADSET ||
-               devices & AUDIO_DEVICE_OUT_DGTL_DOCK_HEADSET) {
+        snd_device = SND_DEVICE_OUT_HDMI;
+    } else if (devices & AUDIO_DEVICE_OUT_ANLG_DOCK_HEADSET) {
+        ALOGD("%s: setting USB hadset channel capability(2) for Proxy", __func__);
+        audio_extn_set_afe_proxy_channel_mixer(adev, 2);
+        snd_device = SND_DEVICE_OUT_USB_HEADSET;
+    } else if (devices & AUDIO_DEVICE_OUT_DGTL_DOCK_HEADSET) {
+#ifndef ASUS_PADSTATION
         ALOGD("%s: setting USB hadset channel capability(2) for Proxy", __func__);
         audio_extn_set_afe_proxy_channel_mixer(adev, 2);
         snd_device = SND_DEVICE_OUT_USB_HEADSET;
+#else
+        ALOGD("%s: enable Asus PAD Speaker", __func__);
+        snd_device = SND_DEVICE_OUT_PAD_SPEAKER;
+#endif
 #ifdef FM_ENABLED
     } else if (devices & AUDIO_DEVICE_OUT_FM_TX) {
         snd_device = SND_DEVICE_OUT_TRANSMISSION_FM;
diff --git a/hal/msm8974/platform.h b/hal/msm8974/platform.h
index 8c019fe..aa151e3 100644
--- a/hal/msm8974/platform.h
+++ b/hal/msm8974/platform.h
@@ -58,9 +58,15 @@ enum {
     SND_DEVICE_OUT_BEGIN = SND_DEVICE_MIN,
     SND_DEVICE_OUT_HANDSET = SND_DEVICE_OUT_BEGIN,
     SND_DEVICE_OUT_SPEAKER,
+#ifdef ASUS_PADSTATION
+    SND_DEVICE_OUT_PAD_SPEAKER,
+#endif
     SND_DEVICE_OUT_SPEAKER_REVERSE,
     SND_DEVICE_OUT_HEADPHONES,
     SND_DEVICE_OUT_SPEAKER_AND_HEADPHONES,
+#ifdef ASUS_PADSTATION
+    SND_DEVICE_OUT_PAD_SPEAKER_AND_HEADPHONES,
+#endif
     SND_DEVICE_OUT_VOICE_HANDSET,
     SND_DEVICE_OUT_VOICE_SPEAKER,
     SND_DEVICE_OUT_VOICE_HEADPHONES,
